libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -3643,3 +3679,262 @@ instance HasSpec fn a => Pretty (Spec fn a) where
 
 instance HasSpec fn a => Show (Spec fn a) where
   showsPrec d = shows . pretty . WithPrec d
+
+-- ==================================================================
+-- Size, Sized, and SizeSpec
+-- ==================================================================
+
+newtype Size = Size {unSize :: Int} deriving (Eq, Ord, Num)
+
+instance Show Size where show (Size n) = "(Size " ++ show n ++ ")"
+
+instance {-# INCOHERENT #-} BaseUniverse fn => OrdLike fn Size where
+  leqSpec (Size n) = typeSpec (SizeSpec (NumSpecInterval Nothing (Just n)))
+  geqSpec (Size n) = typeSpec (SizeSpec (NumSpecInterval (Just n) Nothing))
+  gtSpec (Size n) = typeSpec (SizeSpec (NumSpecInterval (Just (n + 1)) Nothing))
+  ltSpec (Size 0) = ErrorSpec ["Can't make a ltSpec for (Size 0)"]
+  ltSpec (Size n) = typeSpec (SizeSpec (NumSpecInterval (Just (n - 1)) Nothing))
+
+conformsToSize :: Size -> SizeSpec -> Bool
+conformsToSize (Size x) (SizeSpec numspec) = conformsToNumSpec x numspec
+
+instance BaseUniverse fn => HasSpec fn Size where
+  type TypeSpec fn Size = SizeSpec
+  type Prerequisites fn Size = ()
+  emptySpec = SizeSpec mempty
+  combineSpec s1 s2 = TypeSpec (s1 <> s2) []
+  genFromTypeSpec = genFromSizeSpec deltaMax
+  conformsTo = conformsToSize
+  toPreds x szspec@(SizeSpec _) = toPredsSizeSpec x szspec
+  sizeOfTypeSpec x = x
+
+instance {-# INCOHERENT #-} BaseUniverse fn => Num (Term fn Size) where
```

Comment:
This instance suggests that perhaps it would be better for the size functions to just return an `Int`? It would be easier to work with, produce less complex code, and look more like native Haskell.